toolchain/check/impl_lookup.cpp

```diff
@@ -347,6 +349,164 @@ static auto LookupImplWitnessInSelfFacetValue(
   return EvalImplLookupResult::MakeNonFinal();
 }
 
+// Substitutes witnesess in place of `LookupImplWitness` queries into `.Self`,
+// when the witness is for the same interface as the one `.Self` is referring
+// to.
+//
+// This allows access to the `FacetType` and its constraints from the witness,
+// and allows `ImplWitnessAccess` instructions to be immediately resolved to a
+// more specific value when possible.
+class SubstWitnessesCallbacks : public SubstInstCallbacks {
+ public:
+  // `context` must not be null.
+  explicit SubstWitnessesCallbacks(
+      Context* context, SemIR::LocId loc_id,
+      llvm::ArrayRef<SemIR::SpecificInterface> interfaces,
+      llvm::ArrayRef<SemIR::InstId> witness_inst_ids)
+      : SubstInstCallbacks(context),
+        loc_id_(loc_id),
+        interfaces_(interfaces),
+        witness_inst_ids_(witness_inst_ids) {}
+
+  auto Subst(SemIR::InstId& inst_id) -> SubstResult override {
+    // `FacetType` can be concrete even when it has rewrite constraints that
+    // have a symbolic dependency on `.Self`. See use of
+    // `GetConstantValueIgnoringPeriodSelf` in eval. So in order to recurse into
+    // `FacetType` we must check for it before the `is_concrete` early return.
+    if (context().insts().Is<SemIR::FacetType>(inst_id)) {
+      ++facet_type_depth_;
+      return SubstOperands;
+    }
+
+    if (context().constant_values().Get(inst_id).is_concrete()) {
+      return FullySubstituted;
+    }
+
+    auto access = context().insts().TryGetAs<SemIR::ImplWitnessAccess>(inst_id);
+    if (!access) {
+      return SubstOperands;
+    }
+
+    auto lookup =
+        context().insts().GetAs<SemIR::LookupImplWitness>(access->witness_id);
```

Comment:
Is it possible for the impl witness to be of a different kind here? Or must the impl witness of a symbolic `ImplWitnessAccess` constant always be a `LookupImplWitness`?